scripts/nlp_language_modeling/build_retrieval_index.py

```diff
@@ -191,31 +350,63 @@ def get_emb():
 
     # get first batch of sentences to build up the index
     # sentences = get_sentence_chunks()
-
-    emb = get_emb()
-
-    nlist = 100
-    # m is number of subquantizers. So vector of size D is broken into m sub-vectors of size D/m
-    m = args.subquantizers
-    k = 4  # num_nearest neighbors to get
-    quantizer = faiss.IndexFlatIP(emb.shape[1])
-    index = faiss.IndexIVFPQ(quantizer, emb.shape[1], nlist, m, 8)
-    # 8 specifies that each sub-vector is encoded as 8 bits
-    # build the index
-    index.train(emb)
-    logging.info('Trained Index')
-
-    # add the first batch to the index
-    index.add(emb)
+    if args.stage is None or args.stage == 0:
+        emb, slice_id = get_emb()
+        # initialize the Faiss index
+        # m is number of subquantizers. So vector of size D is broken into m sub-vectors of size D/m
+        m = args.subquantizers
+        k = 4  # num_nearest neighbors to get
+        quantizer = faiss.IndexFlatIP(emb.shape[1])
+        # 8 specifies that each sub-vector is encoded as 8 bits
+        index = faiss.IndexIVFPQ(quantizer, emb.shape[1], nlist, m, 8)
```

Comment:
@yidong72 I think it's worth looking into other indices particularly hnsw as well. In NMT it didn't make much of a difference but I am not sure in RETRO how much impact the quality of NNs would have. My guess would be quite a lot so some analysis on improving nearest neighbor quality we can look at.

Reply:
Yeah fair point. GPU acceleration and sharding are important.